Initial work towards optimizing FileService handlers

Therzok opened the pull request:
Fixes VSTS #853896 - [Feedback] Visual Studio freezes when i run mono
Fixes VSTS #902435 - Investigate performance issue with FileSystemWatcher

Therzok committed:
Initial work towards optimizing FileService handlers

Fixes VSTS #853896 - [Feedback] Visual Studio freezes when i run mono
Fixes VSTS #902435 - Investigate performance issue with FileSystemWatcher

Therzok committed:
Add pretty printing to PathTree

Therzok committed:
Optimize glob checking code

Therzok committed:
Increase the default number of file watchers on macos for finer grained watching

Therzok committed:
Optimize some file creation code

Therzok committed:
[Core] O(n) hashcode + O(logn) lookup is way more expensive than the shortcircuit exit from IsChildPathOf in most cases

Therzok committed:
Optimize TypeSystemService handlers

Current timings:
Source MonoDevelop.Core.AsyncEvents
00:03:05.2276958 - MonoDevelop.Projects.Project Void OnFileCreated(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:26.3610085 - MonoDevelop.Projects.Project Void OnFileDeleted(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.0490641 - MonoDevelop.Projects.Project Void OnFileRenamed(System.Object, MonoDevelop.Core.FileCopyEventArgs)
Source MonoDevelop.Core.EventQueue
00:01:47.1092361 - MonoDevelop.Projects.Project Void HandleFileChanged(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:13.1584333 - MonoDevelop.Ide.TypeSystem.MonoDevelopMetadataReference Void OnUpdatedOnDisk(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:10.5654575 - MonoDevelop.Projects.FileStatusTracker`1[MonoDevelop.Projects.SolutionItemEventArgs] Void HandleFileChanged(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:07.8287647 - MonoDevelop.Ide.Gui.Pads.ProjectPad.ShowAllFilesBuilderExtension Void OnSystemFileAdded(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:07.1494449 - MonoDevelop.Ide.TypeSystem.TypeSystemService Void FileService_FileChanged(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:04.0628475 - MonoDevelop.Projects.DotNetProject Void OnFileRemoved(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:03.4642978 - MonoDevelop.Ide.Gui.Pads.ProjectPad.ShowAllFilesBuilderExtension Void OnSystemFileDeleted(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:01.9202754 - MonoDevelop.Unity.Project.NodeBuilders.UnityFileNodeBuilder Void OnFileChanged(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:01.3188652 - MonoDevelop.Ide.RootWorkspace Void CheckFileRemoved(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.7426572 - MonoDevelop.PackageManagement.SdkProjectReloadMonitor Void FileChanged(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.5195917 - MonoDevelop.PackageManagement.NuGetConfigFileChangedMonitor Void FileChanged(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.3707801 - MonoDevelop.Ide.Gui.Pads.ProjectPad.ProjectNodeBuilder Void OnSystemFileCreated(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.3162080 - MonoDevelop.Projects.FileStatusTracker`1[MonoDevelop.Projects.WorkspaceItemEventArgs] Void HandleFileChanged(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.3035524 - MonoDevelop.PackageManagement.NodeBuilders.ProjectPackagesProjectNodeBuilderExtension Void FileChanged(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.2696685 - MonoDevelop.Ide.Gui.DocumentRegistry Void HandleFileServiceChange(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.2377755 - MonoDevelop.Ide.TypeSystem.MonoDevelopAnalyzer Void OnUpdatedOnDisk(System.Object, System.EventArgs)
00:00:00.1657999 - MonoDevelop.PackageManagement.Commands.PackageManagementStartupHandler Void FileChanged(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.0859679 - MonoDevelop.Ide.DesktopService Void NotifyFileRemoved(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.0716940 - MonoDevelop.Ide.Gui.Documents.DocumentManager Void CheckRemovedFile(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.0414073 - MonoDevelop.Ide.Editor.EditorConfigService+ConventionsFileManager Void FileService_FileChanged(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.0195178 - MonoDevelop.Unity.Project.NodeBuilders.UnityFileNodeBuilder Void OnFileCopied(System.Object, MonoDevelop.Core.FileCopyEventArgs)
00:00:00.0026554 - MonoDevelop.Ide.Gui.Pads.ProjectPad.ProjectNodeBuilder Void OnSystemFileDeleted(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.0022615 - MonoDevelop.Ide.Editor.EditorConfigService+ConventionsFileManager Void FileService_FileRemoved(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.0019598 - MonoDevelop.Ide.Gui.Pads.ProjectPad.ShowAllFilesBuilderExtension Void OnSystemFileRenamed(System.Object, MonoDevelop.Core.FileCopyEventArgs)
00:00:00.0018584 - MonoDevelop.Debugger.DebuggingService Void FileService_FileRenamed(System.Object, MonoDevelop.Core.FileCopyEventArgs)
00:00:00.0011180 - MonoDevelop.Ide.Gui.Documents.DocumentManager Void CheckRenamedFile(System.Object, MonoDevelop.Core.FileCopyEventArgs)
00:00:00.0011009 - MonoDevelop.Ide.Gui.Pads.ProjectPad.ProjectFolder Void OnFileRenamed(System.Object, MonoDevelop.Core.FileCopyEventArgs)
00:00:00.0007078 - MonoDevelop.Ide.Editor.EditorConfigService+ConventionsFileManager Void FileService_FileMoved(System.Object, MonoDevelop.Core.FileCopyEventArgs)
00:00:00.0003805 - MonoDevelop.Ide.RootWorkspace Void CheckFileRename(System.Object, MonoDevelop.Core.FileCopyEventArgs)
00:00:00.0003511 - MonoDevelop.Ide.DesktopService Void NotifyFileRenamed(System.Object, MonoDevelop.Core.FileCopyEventArgs)

Therzok committed:
Some more perf optimizations here

```
Source MonoDevelop.Core.AsyncEvents
00:02:34.7219796 - MonoDevelop.Projects.Project Void OnFileCreated(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:01.5497746 - MonoDevelop.Projects.Project Void OnFileDeleted(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.0229513 - MonoDevelop.Projects.Project Void OnFileRenamed(System.Object, MonoDevelop.Core.FileCopyEventArgs)
Source MonoDevelop.Core.EventQueue
00:02:10.0724091 - MonoDevelop.Projects.Project Void HandleFileChanged(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:18.8409343 - MonoDevelop.Ide.TypeSystem.MonoDevelopMetadataReference Void OnUpdatedOnDisk(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:15.7531517 - MonoDevelop.Projects.FileStatusTracker`1[MonoDevelop.Projects.SolutionItemEventArgs] Void HandleFileChanged(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:11.4172185 - MonoDevelop.Ide.Gui.Pads.ProjectPad.ShowAllFilesBuilderExtension Void OnSystemFileAdded(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:10.6727282 - MonoDevelop.Ide.TypeSystem.TypeSystemService Void FileService_FileChanged(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:02.6945364 - MonoDevelop.Unity.Project.NodeBuilders.UnityFileNodeBuilder Void OnFileChanged(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:01.6960784 - MonoDevelop.Projects.FileStatusTracker`1[MonoDevelop.Projects.WorkspaceItemEventArgs] Void HandleFileChanged(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:01.0363946 - MonoDevelop.PackageManagement.SdkProjectReloadMonitor Void FileChanged(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.5363695 - MonoDevelop.PackageManagement.NuGetConfigFileChangedMonitor Void FileChanged(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.4974428 - MonoDevelop.Ide.Gui.Pads.ProjectPad.ProjectNodeBuilder Void OnSystemFileCreated(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.4228951 - MonoDevelop.PackageManagement.NodeBuilders.ProjectPackagesProjectNodeBuilderExtension Void FileChanged(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.4220673 - MonoDevelop.Ide.Gui.DocumentRegistry Void HandleFileServiceChange(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.4020246 - MonoDevelop.PackageManagement.Commands.PackageManagementStartupHandler Void FileChanged(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.3517315 - MonoDevelop.Ide.TypeSystem.MonoDevelopAnalyzer Void OnUpdatedOnDisk(System.Object, System.EventArgs)
00:00:00.2972448 - MonoDevelop.Ide.Gui.Pads.ProjectPad.ShowAllFilesBuilderExtension Void OnSystemFileDeleted(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.1205992 - MonoDevelop.Ide.RootWorkspace Void CheckFileRemoved(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.0819609 - MonoDevelop.Projects.DotNetProject Void OnFileRemoved(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.0697519 - MonoDevelop.Ide.Editor.EditorConfigService+ConventionsFileManager Void FileService_FileChanged(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.0325364 - MonoDevelop.Unity.Project.NodeBuilders.UnityFileNodeBuilder Void OnFileCopied(System.Object, MonoDevelop.Core.FileCopyEventArgs)
00:00:00.0199356 - MonoDevelop.Ide.DesktopService Void NotifyFileRemoved(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.0117852 - MonoDevelop.Ide.Gui.Documents.DocumentManager Void CheckRemovedFile(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.0025223 - MonoDevelop.Ide.Gui.Pads.ProjectPad.ProjectNodeBuilder Void OnSystemFileDeleted(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.0024015 - MonoDevelop.Ide.Gui.Pads.ProjectPad.ShowAllFilesBuilderExtension Void OnSystemFileRenamed(System.Object, MonoDevelop.Core.FileCopyEventArgs)
00:00:00.0016862 - MonoDevelop.Ide.Editor.EditorConfigService+ConventionsFileManager Void FileService_FileRemoved(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.0012199 - MonoDevelop.Ide.Gui.Pads.ProjectPad.ProjectFolder Void OnFileRenamed(System.Object, MonoDevelop.Core.FileCopyEventArgs)
00:00:00.0010056 - MonoDevelop.Debugger.DebuggingService Void FileService_FileRenamed(System.Object, MonoDevelop.Core.FileCopyEventArgs)
00:00:00.0009376 - MonoDevelop.Ide.Gui.Documents.DocumentManager Void CheckRenamedFile(System.Object, MonoDevelop.Core.FileCopyEventArgs)
00:00:00.0006533 - MonoDevelop.Ide.Editor.EditorConfigService+ConventionsFileManager Void FileService_FileMoved(System.Object, MonoDevelop.Core.FileCopyEventArgs)
00:00:00.0004185 - MonoDevelop.Ide.DesktopService Void NotifyFileRenamed(System.Object, MonoDevelop.Core.FileCopyEventArgs)
00:00:00.0003588 - MonoDevelop.Ide.RootWorkspace Void CheckFileRename(System.Object, MonoDevelop.Core.FileCopyEventArgs)
```

Therzok committed:
Switch back to normal invoke calls

Therzok commented on main/src/core/MonoDevelop.Core/CoreExtensions.cs:
```diff
@@ -296,6 +297,82 @@ public static void SafeInvoke (this EventHandler events, object sender, EventArg
 			}
 		}
+
+		internal static void TimeInvoke<T> (this EventHandler<T> events, object sender, T args)
```
Left this code in for future use.

Therzok commented:
The core problem remains unsolved still. We need to figure out a mechanism where we don't have all the files notified on all the handlers.

This was the initial state: https://gist.github.com/Therzok/33c8b7d973257cee558ac11cf1551f0d
This is the current one: https://gist.github.com/Therzok/fd9c7f4d11a1483947169daa9b9f2eee

While there is an improvement, we're still calling into event handlers _way_ too often. We need some kind of prefiltering to be done and maybe not watching every directory. Most of the work done here is to mitigate the symptoms.

There is no easy fix we can make to actually improve the general situation.

Therzok committed:
Move PathTreeNode pretty printing into pathtreenode

mrward commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects/Project.cs:
```diff
@@ -4488,17 +4498,17 @@ void DisposeFileWatcher ()
 			foreach (FileEventInfo info in e) {
-				OnFileRenamed (info.SourceFile, info.TargetFile);
+				OnFileRenamed (info.SourceFile, info.TargetFile, info.IsDirectory);
 			}
```
Not sure this going to work in all cases. We only set the IsDirectory when the user renames a directory. The file watcher events do not indicate this.

mrward commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects/Project.cs:
```diff
@@ -4518,25 +4528,30 @@ void OnFileCreated (object sender, FileEventArgs e)
 			foreach (FileEventInfo info in e) {
-				OnFileCreated (info.FileName);
+				OnFileCreated (info.FileName, info.IsDirectory);
 			}
```
Similar here - this will be false for file watcher generated events.

mrward commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects/Project.cs:
```diff
@@ -4551,7 +4566,7 @@ void OnFileDeleted (object sender, FileEventArgs e)
 				foreach (FileEventInfo info in e) {
-					OnFileDeletedExternally (info.FileName);
+					OnFileDeletedExternally (info.FileName, info.IsDirectory);
 				}
```
This will always be false for file watcher events. It is not possible to know if a directory is deleted with the file watcher since it is gone before we can check.

mrward commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects/Project.cs:
```diff
@@ -4633,34 +4647,38 @@ void OnFileCreatedExternally (FilePath fileName)
+			bool notDeleted = isDirectory
+				? Directory.Exists (fileName)
+				: File.Exists (fileName);
+
```
I do not think this is going to work. We cannot rely on the isDirectory flag for the file watcher generated events.

mrward reviewed (commented):
Most of this looks good to me. Just some potential problems with the use of the IsDirectory flag.

Therzok commented:
Test failures are related to the Exists checks. Will revert that and get newer numbers.

Therzok committed:
[Core] Revert existence checks on FSW notifications

We can't rely on info.IsDirectory until https://github.com/mono/monodevelop/issues/7963 is fixed

Therzok committed:
[Core] Speed up project file removal by doing a cheap O(logn) lookup before iterating the project file list

Therzok committed:
Cleanup some code in FileSystemWatcher

```
Source MonoDevelop.Core.AsyncEvents
00:03:01.2376095 - MonoDevelop.Projects.Project Void OnFileCreated(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:14.4988996 - MonoDevelop.Projects.Project Void OnFileDeleted(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.0258500 - MonoDevelop.Projects.Project Void OnFileRenamed(System.Object, MonoDevelop.Core.FileCopyEventArgs)
Source MonoDevelop.Core.EventQueue
00:00:53.8239910 - MonoDevelop.Projects.Project Void HandleFileChanged(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:05.9432093 - MonoDevelop.Ide.TypeSystem.MonoDevelopMetadataReference Void OnUpdatedOnDisk(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:05.6516870 - MonoDevelop.Ide.Gui.Pads.ProjectPad.ShowAllFilesBuilderExtension Void OnSystemFileAdded(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:03.8726403 - MonoDevelop.Projects.FileStatusTracker`1[MonoDevelop.Projects.SolutionItemEventArgs] Void HandleFileChanged(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:02.1907880 - MonoDevelop.Ide.Gui.Pads.ProjectPad.ShowAllFilesBuilderExtension Void OnSystemFileDeleted(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:01.5677551 - MonoDevelop.Ide.TypeSystem.TypeSystemService Void FileService_FileChanged(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:01.4997610 - MonoDevelop.Unity.Project.NodeBuilders.UnityFileNodeBuilder Void OnFileChanged(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.8444511 - MonoDevelop.Ide.RootWorkspace Void CheckFileRemoved(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.3326067 - MonoDevelop.Ide.Gui.Pads.ProjectPad.ProjectNodeBuilder Void OnSystemFileCreated(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.2527089 - MonoDevelop.Projects.DotNetProject Void OnFileRemoved(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.2517709 - MonoDevelop.PackageManagement.SdkProjectReloadMonitor Void FileChanged(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.1316798 - MonoDevelop.PackageManagement.NuGetConfigFileChangedMonitor Void FileChanged(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.1239526 - MonoDevelop.PackageManagement.NodeBuilders.ProjectPackagesProjectNodeBuilderExtension Void FileChanged(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.1091269 - MonoDevelop.Projects.FileStatusTracker`1[MonoDevelop.Projects.WorkspaceItemEventArgs] Void HandleFileChanged(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.0940398 - MonoDevelop.Ide.TypeSystem.MonoDevelopAnalyzer Void OnUpdatedOnDisk(System.Object, System.EventArgs)
00:00:00.0870917 - MonoDevelop.PackageManagement.Commands.PackageManagementStartupHandler Void FileChanged(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.0468959 - MonoDevelop.Ide.DesktopService Void NotifyFileRemoved(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.0326173 - MonoDevelop.Unity.Project.NodeBuilders.UnityFileNodeBuilder Void OnFileCopied(System.Object, MonoDevelop.Core.FileCopyEventArgs)
00:00:00.0039422 - MonoDevelop.Ide.Gui.Documents.DocumentManager Void CheckRemovedFile(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.0035815 - MonoDevelop.Debugger.DebuggingService Void FileService_FileRenamed(System.Object, MonoDevelop.Core.FileCopyEventArgs)
00:00:00.0015879 - MonoDevelop.Ide.Gui.Pads.ProjectPad.ShowAllFilesBuilderExtension Void OnSystemFileRenamed(System.Object, MonoDevelop.Core.FileCopyEventArgs)
00:00:00.0010635 - MonoDevelop.Ide.Gui.Pads.ProjectPad.ProjectFolder Void OnFileRenamed(System.Object, MonoDevelop.Core.FileCopyEventArgs)
00:00:00.0009276 - MonoDevelop.Ide.Gui.Pads.ProjectPad.ProjectNodeBuilder Void OnSystemFileDeleted(System.Object, MonoDevelop.Core.FileEventArgs)
00:00:00.0007076 - MonoDevelop.Ide.Gui.Documents.DocumentManager Void CheckRenamedFile(System.Object, MonoDevelop.Core.FileCopyEventArgs)
00:00:00.0003854 - MonoDevelop.Ide.DesktopService Void NotifyFileRenamed(System.Object, MonoDevelop.Core.FileCopyEventArgs)
00:00:00.0003510 - MonoDevelop.Ide.RootWorkspace Void CheckFileRename(System.Object, MonoDevelop.Core.FileCopyEventArgs)
```

Therzok committed:
[Core] Refactor fileservice event queue handling code so it's testable

Therzok committed:
[Core] Avoid reflection in AsyncEvents

Therzok committed:
Add counter IDs to file service counters

Therzok committed:
[AutoTest] Add some helper methods to validate counter hit count

Therzok committed:
Add AsyncEvents timings

Therzok committed:
[Core] Add Performance tests for file watcher changes

Therzok committed:
Add sdk library for testing performance issues with watchers

Therzok committed:
[Core] Fix unit tests. The list needs to be cleared if it's to be reused

Therzok committed:
[Core] Refactor a bit of list pooling to optimize for common case

Therzok committed:
Move existence checks for when a file is deleted onto the background thread

Therzok commented:
Quite a few more changes. Added unit tests and perf tests for validating data spent in the file watcher.

Therzok commented:
Only need to get a new baseline for the perf tests. Not sure why the others are faster now too?

slluis commented:
@monojenkins backport release-8.2

Therzok committed:
Update new baseline

Therzok committed:
Log on CI which directory it failed to cleanup

Therzok commented:
Perf tests failed. A new baseline is needed. I"d rather merge this as is, and fix it up asap

Therzok commented:
Ah, no, they didn't

Therzok commented:
@monojenkins backport release-8.2

Therzok commented:
@slluis pending your approval

Therzok commented:
I will make a new batch of perf fixes starting from this PR.

slluis commented:
@monojenkins backport release-8.2
